linkerd/http/retry/src/peek_trailers.rs

```diff
@@ -17,31 +18,34 @@ use std::{
 /// If the first frame of the body stream was *not* a `TRAILERS` frame, this
 /// behaves identically to a normal body.
 #[pin_project(project = Projection)]
-pub struct PeekTrailersBody<B: Body = BoxBody> {
-    /// The inner [`Body`].
+pub enum PeekTrailersBody<B: Body = BoxBody> {
+    /// An empty body.
+    Empty,
+    /// A body that contains zero or one DATA frame.
     ///
-    /// This is the request or response body whose trailers are being peeked.
-    #[pin]
-    inner: B,
-
-    /// The first DATA frame received from the inner body, or an error that
-    /// occurred while polling for data.
+    /// This variant MAY have trailers that can be peeked.
+    Unary {
+        data: Option<Result<B::Data, B::Error>>,
+        trailers: Option<Result<HeaderMap, B::Error>>,
+    },
+    /// A body that (potentially) contains more than one DATA frame.
     ///
-    /// If this is `None`, then the body has completed without any DATA frames.
-    first_data: Option<Result<B::Data, B::Error>>,
-
-    /// The inner body's trailers, if it was terminated by a `TRAILERS` frame
-    /// after 0-1 DATA frames, or an error if polling for trailers failed.
+    /// This variant indicates that the inner body's trailers could not be observed.
+    Unknown {
+        first: Option<Result<B::Data, B::Error>>,
+        second: Option<Result<B::Data, B::Error>>,
+        /// The inner [`Body`].
+        #[pin]
+        inner: B,
+    },
```

Comment:
bikeshedding for this variant is welcome. `Stream` seemed like another alternative, but it feels like a particularly overloaded term. i chose `Unknown` because it doesn't _guarantee_ that two DATA frames exist: the second frame may have not been immediately available.

Reply:
Perhaps `Buffered`? In this state we've buffered data and can't peek further.

Reply:
i like this name much better, thank you! renamed in 2557a2f3